File: 2024 January/Search Pattern (KMP-Algorithm)/main.py
# Link: https://www.geeksforgeeks.org/problems/search-pattern0205/1


class Solution:
    def search(self, pat, txt):
        indices = []
        n = len(txt)
        m = len(pat)

        for i in range(n - m + 1):
            j = 0
            while j < m:
                if txt[i + j] != pat[j]:
                    break
                j += 1

            if j == m:
                indices.append(i + 1)

        if len(indices) == 0:
            return [-1]
        else:
            return indices


if __name__ == "__main__":
    obj = Solution()
    txt = "geeksforgeeks"
    pat = "geek"
    ans = obj.search(pat, txt)
    for i in ans:
        print(i, end=" ")
    print()

    txt = "abesdu"
    pat = "edu"
    ans = obj.search(pat, txt)
    for i in ans:
        print(i, end=" ")

File: 2024 January/Search Pattern (KMP-Algorithm)/main.cpp
// Link: https://www.geeksforgeeks.org/problems/search-pattern0205/1

#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    vector<int> search(string pat, string txt)
    {
        vector<int> indices;
        int n = txt.length();
        int m = pat.length();

        for (int i = 0; i <= n - m; ++i)
        {
            int j;
            for (j = 0; j < m; ++j)
            {
                if (txt[i + j] != pat[j])
                {
                    break;
                }
            }

            if (j == m)
            {
                indices.push_back(i + 1);
            }
        }
        return indices;
    }
};

int main()
{
    Solution obj;
    string txt = "abesdu";
    string pat = "edu";
    vector<int> ans = obj.search(pat, txt);
    for (int i = 0; i < ans.size(); i++)
    {
        cout << ans[i] << " ";
    }
    cout << endl;

    string txt1 = "geeksforgeeks";
    string pat1 = "geek";
    vector<int> ans1 = obj.search(pat1, txt1);
    for (int i = 0; i < ans1.size(); i++)
    {
        cout << ans1[i] << " ";
    }
    cout << endl;
    return 0;
}